### ISSUE TITLE
gateleen-delegate: allow payload transformation

### ISSUE THREAD

**Issue by mcweba** [swisspush/gateleen#174]
The current implementation of the delegate feature allows to define a payload for the delegate request.

Example delegate definition:
> PUT /gateleen/server/delegate/v1/delegates/greeting/definition
```json
{
    "methods": [ "PUT" ],
    "pattern": ".*/([^/]+.*)",
    "requests": [
        {
            "method": "PUT",
            "uri": "/gateleen/server/resources/greeting",
            "payload": {
                "hello": "world"
            }
        }
    ]
}
```
When the above delegate definition is executed, a PUT request to _/gateleen/server/resources/greeting_ is made with the body
```json
{
    "hello": "world"
}
```
The payload of the original request to the delegate execution
> PUT /gateleen/server/delegate/v1/delegates/greeting/execution

 is simply ignored.

To extend the delegate feature with the ability to work with the payload of the original execution request, the following improvements could be implemented.

## Proposal for solution
Extend the delegate definition with a property called _transform_. This property contains a _Jolt spec_.
The _Jolt_ library is a JSON-to-JSON transformation library written in java. See [bazaarvoice / jolt](https://github.com/bazaarvoice/jolt)

The _Jolt spec_ is a JSON resource defining how to transform the input JSON into the output JSON.

Example spec:
```json
[
  {
    "operation": "shift",
    "spec": {
      "data": {
        "clientId": "id",
        "clientName": "name"
      }
    }
  }
]
```
This spec renames the _clientId_ property to _id_ and the _clientName_ property to _name_ and also move these properties one level up by removing the _data_ property. So for the input
```json
{
  "data": {
    "clientId": "1234",
    "clientName": "Acme"
  }
}
```
the output will be
```json
{
  "id" : "1234",
  "name" : "Acme"
}
```
By introducing the _Jolt spec_ into the new _transform_ property, we could provide a very powerful extension to the delegate feature. Checkout the documentation of _Jolt_ for more examples.

A delegate definition could then look like this:
```json
{
  "methods": ["PUT"],
	"pattern": ".*/([^/]+.*)",
	"requests": [{
		"method": "PUT",
		"uri": "/gateleen/server/resources/greeting",
		"transform": [{
			"operation": "shift",
			"spec": {
				"data": {
					"clientId": "id",
					"clientName": "name"
				}
			}
		}]
	}]
}
```
The _payload_ property and the new _transform_ property cannot be used together. The JSON schema validation of the delegate definition has to check that exactly one of them exists!

Regarding performance, the following issues have to be targeted:

- Jolt is not "stream" based, so if a very large JSON document has to be transformed,  there has to be enough memory to hold it
- JSON transformation could take quite some time based on the size of the JSON document to transform and the complexity of the spec (using lots of wildcard logic), therefore it could make sense to implement the transformation in a [_executeBlocking_](http://vertx.io/docs/apidocs/io/vertx/core/Vertx.html#executeBlocking-io.vertx.core.Handler-boolean-io.vertx.core.Handler-) block

**Comment by lbovet** [comment]
1. Better use simply the property "transform" (underscore are forbidden in JSON property names, btw).
2. Is jolt able to map the whole incoming payload to an outgoing property?
3. Actually, what woul dbe the Jolt spec for the identity function (outgoing payload equals incoming payload)?

**Comment by mcweba** [comment]
1. I changed the property to transform (thx for the hint about underscores)
2. Yes this is possible. The spec for the currently known use case would be very simple:
```json
[
  {
    "operation": "shift",
    "spec": {
      "@1": {
        "@": "records[].value"
      }
    }
  }
]
```
3. The spec for the identity function is almost the same
```json
[
  {
    "operation": "shift",
    "spec": {
      "@1": {
        "@": ""
      }
    }
  }
]
```
Check out [Jolt Transform Demo](http://jolt-demo.appspot.com/#inception) to play with the specs

**Comment by lbovet** [comment]
Looks good!

**Comment by mcweba** [comment]
Released in [v1.1.9](https://github.com/swisspush/gateleen/releases/tag/v1.1.9)
